compiler/rustc_typeck/src/check/wfcheck.rs

```diff
@@ -200,6 +200,59 @@ pub fn check_trait_item(tcx: TyCtxt<'_>, def_id: LocalDefId) {
     };
     check_object_unsafe_self_trait_by_name(tcx, &trait_item);
     check_associated_item(tcx, trait_item.hir_id(), span, method_sig);
+
+    let encl_trait_hir_id = tcx.hir().get_parent_item(hir_id);
+    let encl_trait = tcx.hir().expect_item(encl_trait_hir_id);
+    let encl_trait_def_id = encl_trait.def_id.to_def_id();
+    let fn_lang_item_name = if Some(encl_trait_def_id) == tcx.lang_items().fn_trait() {
+        Some("fn")
+    } else if Some(encl_trait_def_id) == tcx.lang_items().fn_mut_trait() {
+        Some("fn_mut")
+    } else {
+        None
+    };
+
+    if fn_lang_item_name.is_some() && trait_item.ident.name.to_ident_string() == "call" {
+        let fn_lang_item_name = fn_lang_item_name.unwrap();
```

Comment:
I think the following should typecheck and do the same without the need for `unwrap()`. It's not a big deal here, but it's generally a good idea to avoid it because code can change over time and the check that ensured the `unwrap()` doesn't panic could go away, while if you're relying on destructuring that can never happen.

```suggestion
    if let (Some(fn_lang_item_name), "call") = (fn_lang_item_name, trait_item.ident.name.to_ident_string().as_str()) {
```

Reply:
Fixed. Thanks for the suggestion.